Faster float parsing

michaeleisel opened the pull request:
@lemire here's a version that works for all floats in canada.json. there are a few edge cases to be worked out, and the perf does take a hit (about a 20% increase). but i will be playing with using a mix of powers of 10 and 5, or just using both all 10 and all 5, to see if i can get more accurate mantissas (or at least, mantissas whose inaccuracy is not correlated). it does hit the fast path most of the time, just gets hit hard by what i guess is a slow strtod on this mac

michaeleisel commented:
It looks pretty good to me now (two bugs fixed). I think the math is very sound, I just need to add handling for +-inf or nan or w/e. And to fix the unit test fail. However, when I run the unit tests locally and it tells me the two temp files don't match, they seem to be gone afterward when I try to open them. So it's hard for me to debug, and I'm not sure if the issue is actually related to my PR.

The perf is better now. It's about 15% slower for canada.json than the current way. It hits the fast path about 99 times out of 100 for that file. even so, strtod is pretty slow on my mac when it hits the slow path (10x slower). Some things to do from here:
- throw this in (we can add support for when `i` > 2 ** 64 as well, it will still hit the fast path a lot) and just let it be slow occasionally
- use a different fallback instead of strtod. there are lots, e.g. the bigInt one that rapidjson does, or use a 128-bit mantissa instead of 64 bits and do the same algo

michaeleisel commented:
we can also still use the old way in some situations of course when it will do full precision

lemire commented:
Fantastic.

I'll want to review this line by line but I am optimistic.

We want to maintain fast parsing...

michaeleisel commented:
yeah, i just cleaned it up in case you're on an older commit. i see now that the test fails for a large value, that will be fixed. i think this float parsing may be novel, if so i may want to publish it. do you know how i would do due diligence on seeing if anyone's published?

michaeleisel commented:
also, i'm happy to add explanations on the pr/in the comments if we definitely want to go through with it. the hacks are a bit unapproachable

lemire commented on include/simdjson/numberparsing.h:
```diff
@@ -562,9 +572,33 @@ static really_inline bool parse_number(const uint8_t *const buf, ParsedJson &pj,
+        uint64_t real_exponent = c.exp + 1023 + (127 - lz);
+        mantissa |= real_exponent << 52;
+        mantissa |= ((uint64_t)negative) << 63;
+        d = *((double *)&mantissa);
```
This is almost certainly not 'safe'. I recommend doing... `memcpy(&d,&mantissa,8)`.

michaeleisel commented on include/simdjson/numberparsing.h:
```diff
@@ -562,9 +572,33 @@ static really_inline bool parse_number(const uint8_t *const buf, ParsedJson &pj,
+      uint64_t factor_mantissa = c.mantissa;
+      int lz = leading_zeroes(i);
+      i <<= lz;
+      __uint128_t large_mantissa = (__uint128_t)i * factor_mantissa;
```
128-bit support is an open question... rapidjson, for example, has a fallback when it detects that it's not possible (see their bigint.h file). it seems pretty common to support it but idk

lemire replied:
We need to support Visual Studio...

michaeleisel replied:
ok that's fine, we can just compute this without it (3 multiplications and a few adds and shifts)

lemire commented on include/simdjson/numberparsing.h:
```diff
@@ -562,9 +572,33 @@ static really_inline bool parse_number(const uint8_t *const buf, ParsedJson &pj,
+        mantissa |= ((uint64_t)negative) << 63;
+        d = *((double *)&mantissa);
+      } else {
+        d = strtod((char *)(buf + offset), NULL);
```
The bulk of the performance loss is due to this call, and we don't even have perfect parsing (e.g., elsewhere we call parse_float).

michaeleisel replied:
hmm what are you saying? btw parse_float isn't called in canada.json

lemire commented:
@michaeleisel 

> hmm what are you saying? btw parse_float isn't called in canada.json

The strtod calls are super expensive (they dominate the running time even if they are rarely called). And this expense does not even buy you perfect parsing.

Let me submit to you the following...

1. If you are going to have very slow number parsing, then you want to have provably exact parsing. What is the point of very slow number parsing that only happens to be exact on some particular file that may or may not represent what people will get? How do you advertise this feature? `simdjson` has exact parsing, but only on `canada.json`... ?

2. Otherwise, if you don't have exact parsing, then you probably want to fast... as fast as RapidJSON or sajson in their fast parsing mode (ULP = 1).

I am not sure I see the value in "slow but inexact" parsing.

That is not the same thing as not seeing the value of your PR btw...!!! It looks like you have something clever and worthwhile, we just need to work out the objectives here. I don't think users care about one particular file (canada.json).

This is a silly file.

michaeleisel commented:
which floats are you concerned about it being inexact for? it's true that, as-is, we don't support when `i` is too large, which is why this is a rough sketch and why i outlined the options. this method can still support large numbers, we just parse them until we can't fit anymore digits in a 64-bit int and then use that (see the math in my comment "this can also handle when the other term...")

michaeleisel commented:
i apologize if i was not clear about that. maybe i'm misunderstanding something though. anyways, if what i'm saying is correct, we also have a nice fallback for big ints, for the slower path:
i1 = first few digits until we store 64 bits
i2 = next digits until we run out or store 64 bits
we then merge with i1 * (10 ** x) + i2 = I
for mantissas, we have m1 (first 64 bits, precomputed as you see in the PR), m2 (second 64 bits, precomputed and can be added to the PR), and them together as a 128-bit int is M
we can then compute the following, each time checking if it's correct and then moving on the next one if not:
```
i1 * m1
i1 * M
I * m1
I * M
```
and `I * M` will almost always work. in the extremely rare cases it doesn't, something like strtod would be fine

michaeleisel commented:
`m2` and `i2` don't even have to 64 bits, they could be something like 16 bits

lemire commented:
Let me try to be clearer. To my knowledge, RapidJSON has two modes. The default is fast number parsing, like simdjson. It then has a mode for high-precision parsing. 

It makes sense for simdjson to support both modes. So we have the current mode, and then we have a new  mode with high precision parsing. However, the counterpart to this is that the new mode must be exact (like RapidJSON). And it should be as fast as RapidJSON.

Falling back on the standard library, if you always correctly fall back, buys you perfect parsing... [and note that this means that every code path must have correct fallback] However, you also become slower than RapidJSON (when they are both trying to achieve perfect parsing) on number parsing when the fallback occurs. So I don't think that falling back on the standard library is an option... (because of the performance considerations)

lemire commented:
I think I understand what you are saying: the method can and should be extended when the 64-bit integer can overflow.

I think where we differ is that I'm much less happy than you are about the fallback on the standard library. In my tests, this does not appear to be an "extremely rare" event even just in our test files (to say nothing of the worst case one could encounter in the wild), and it is insanely expensive. A parser like RapidJSON appears to do better. I think we need to put the bar higher in simdjson.

michaeleisel commented:
the code that i have doesn't compute `I * M`, which is the one that should only fail in extremely rare cases. it computes `i1 * m1`, which is much more likely to happen (but still rare). for every bit of precision we add to both `I` and `M`, the chance of needing to use the slow path should drop by 50%. i can implement it and we can see in practice. it's possible that numbers for whatever reason would tend to hit the slow path more than that, but i see no specific reason they would.

if we wanted to remove *any* chance of using strtod, we'd need our own bigint impl like rapidjson does

lemire commented:
Reference: https://docs.microsoft.com/en-us/cpp/intrinsics/umulh

lemire commented:
@michaeleisel 

Ok. Thanks for clarifying.

I think that falling back on the standard library in extreme cases is fine.

michaeleisel commented:
nice, but it's microsoft specific? i don't want to get windows running on my machine. also, it only returns the high bits. i wonder if it's worth it to write out some short assembly functions with the help of godbolt

lemire commented:
> nice, but it's microsoft specific?

It is but the point is that it is possible to support "compute the 64 msb of a 64-bit product" portably and efficiently (without doing the crazy "3 multiplications + adds and shifts" danse.

>  it only returns the high bits

But that's all you are using right now!

On ARM, the high and low bits are distinct instructions... 

>  i wonder if it's worth it to write out some short assembly functions with the help of godbolt

Visual Studio does not support inline assembly.

michaeleisel commented:
what do you think about:
```
if charCount > 19
    success = try64Bits()
    if !success
        try80Bits()  
    else do strtod  
else
    success = try64Bits()
    else do strtod
end
```
so we have the 64 bits method twice. but would we inline? do we care if the same method is inlined by the compiler twice?

lemire commented:
@michaeleisel In our current code (master), we need to branch because we might have an overflow... that's why we have a distinct code path where we reparse the digits. It is uncommon to have so many digits, so branching out to something a bit slower is fine.

Whether one must inline or not is better assessed experimentally, I think. Note that the compiler might not "inline twice". The compiler can optimize away code redundancies.

michaeleisel commented:
btw can we move the found_float and other similar calls to their associated fns, like write_tape_double?

lemire commented:
@michaeleisel These functions are a hack to allow us to have easy tests. They should go away entirely...

michaeleisel commented on include/simdjson/numberparsing.h:
```diff
@@ -572,17 +615,23 @@ static really_inline bool parse_number(const uint8_t *const buf, ParsedJson &pj,
+      return parse_float_strtod(buf, pj, offset, p);
+    }
+    double d = compute_float_double(i, exponent, power_index, negative);
+    if (isnan(d) && true /*high-precision flag here*/) {
```
this is where we'd have a flag for if it should always be precise

michaeleisel commented on include/simdjson/numberparsing.h:
```diff
@@ -572,17 +615,23 @@ static really_inline bool parse_number(const uint8_t *const buf, ParsedJson &pj,
-      return parse_float(buf, pj, offset, found_minus);
+      return parse_float_strtod(buf, pj, offset, p);
+    }
+    double d = compute_float_double(i, exponent, power_index, negative);
```
so this would now do a div, even  if the power is negative. the speed difference didn't seem very big so i figured i'd show this and i can revert if we want

michaeleisel replied:
s/even if/if/

michaeleisel commented:
updated. it doesn't quite pass *all* the tests, and i want to add some unit tests, but this is about what i have in mind. note that i removed all the logic in parse_float, since it seems like we already do all the necessary validation already

michaeleisel commented:
by the way, is there a good way to do perf testing to find perf wins and regressions?

lemire commented:
@michaeleisel  [I think it should be `std::isnan` and not `isnan`](https://en.cppreference.com/w/cpp/numeric/math/isnan).


Your code looks clean and interesting.

Let us talk about performance...

For benchmarks, run `parse` under bare-metal linux... Basically this PR slows things down by 20% to 25%...

Before...

```
$  ./parse jsonexamples/canada.json
number of iterations 10
number of bytes 2251051 number of structural chars 334374 ratio 0.149
mem alloc instructions:       1207 cycles:        987 (0.01 %) ins/cycles: 1.22 mis. branches:          4 (cycles/mis.branch 240.88) cache accesses:     126216 (failure         12)
 mem alloc runs at 0.00 cycles per input byte.
stage 1 instructions:    6469841 cycles:    2001722 (25.37 %) ins/cycles: 3.23 mis. branches:       5570 (cycles/mis.branch 359.36) cache accesses:     126216 (failure       2668)
 stage 1 runs at 0.89 cycles per input byte.
stage 2 instructions:   21645546 cycles:    5887416 (74.62 %) ins/cycles: 3.68 mis. branches:      12019  (cycles/mis.branch 489.82)  cache accesses:     219945 (failure       9942)
 stage 2 runs at 2.62 cycles per input byte and 17.61 cycles per structural character.
 all stages: 3.51 cycles per input byte.
Estimated average frequency: 3.694 GHz.
Min:  0.00213567 bytes read: 2251051 Gigabytes/second: 1.05403
```

After...

```
$  ./parse jsonexamples/canada.json 
number of iterations 10
number of bytes 2251051 number of structural chars 334374 ratio 0.149
mem alloc instructions:       1207 cycles:       1093 (0.01 %) ins/cycles: 1.10 mis. branches:          4 (cycles/mis.branch 273.38) cache accesses:     125837 (failure          8)
 mem alloc runs at 0.00 cycles per input byte.
stage 1 instructions:    6469841 cycles:    1941061 (20.15 %) ins/cycles: 3.33 mis. branches:       5548 (cycles/mis.branch 349.87) cache accesses:     125837 (failure       3231)
 stage 1 runs at 0.86 cycles per input byte.
stage 2 instructions:   27218761 cycles:    7690125 (79.84 %) ins/cycles: 3.54 mis. branches:      14604  (cycles/mis.branch 526.58)  cache accesses:     220379 (failure      10428)
 stage 2 runs at 3.42 cycles per input byte and 23.00 cycles per structural character.
 all stages: 4.28 cycles per input byte.
Estimated average frequency: 3.690 GHz.
Min:  0.00261043 bytes read: 2251051 Gigabytes/second: 0.86233
```

This PR significantly increases the number of instructions in stage 2, it also increases the number of mispredicted branches.

The PR adds about 5573215 instructions, or 2.5 instructions per input byte. Put it another way, `canada.json` contains 111080 floats...  so this PR adds 50 instructions per float in the file.

It would be interesting to know where all these instructions come from. 

In a previous version, almost all of the regression was due to a few calls to the standard library for float parsing (they are ridiculously expensive). I don't think it is the case anymore.

I think that the extra 50 instructions per float come in part from `compute_float_64`. This is frequently called... (maybe 10% of the time) and it is not cheap... 

So the function itself might cost ~ 40 instructions... and because the call to `compute_float_64` is probably hard to predict, you get 15 cycles of penalty right there.

Within  `compute_float_64`, I am concerned with the branch on `justLastBit`. If this is hard to predict, it could cost us another 15 cycles from time to time. I wonder whether it could be made branchless?

michaeleisel commented:
no more branching, and added inlining. leading_zeros could also be replaced with a few adds and shifts, since we can estimate it from the digit count

lemire commented:
The latest changes actually increase the instruction count and lower the performance (slightly). So this looks like a regression but it is a small effect.


```
$ ./parse jsonexamples/canada.json
number of iterations 10
number of bytes 2251051 number of structural chars 334374 ratio 0.149
mem alloc instructions:       1207 cycles:        883 (0.01 %) ins/cycles: 1.37 mis. branches:          3 (cycles/mis.branch 232.50) cache accesses:     125896 (failure          7)
 mem alloc runs at 0.00 cycles per input byte.
stage 1 instructions:    6469841 cycles:    1944649 (20.07 %) ins/cycles: 3.33 mis. branches:       5540 (cycles/mis.branch 350.99) cache accesses:     125896 (failure       2289)
 stage 1 runs at 0.86 cycles per input byte.
stage 2 instructions:   27670063 cycles:    7744437 (79.92 %) ins/cycles: 3.57 mis. branches:      13052  (cycles/mis.branch 593.33)  cache accesses:     219191 (failure       8322)
 stage 2 runs at 3.44 cycles per input byte and 23.16 cycles per structural character.
 all stages: 4.30 cycles per input byte.
Estimated average frequency: 3.692 GHz.
Min:  0.00262425 bytes read: 2251051 Gigabytes/second: 0.857787
```

So my concern was not warranted, you are better with your original code...

```
uint64_t justLastBit = upper & (1ULL << 63);
if (justLastBit) {
mantissa = upper >> 10;
 } else {
mantissa = upper >> 9;
 lz++;
}
```

The reason is unintuitive to me... the savings in instructions exceed the cost of the mispredictions. (Look at how many instructions we save!)

I recommend you revert this change I suggested.

>  leading_zeros could also be replaced with a few adds and shifts, since we can estimate it from the digit count

It is unlikely to be faster if you do it otherwise because modern processors have fast instructions to do it...  Still the whole `lz` is quite expensive but that's because it leads to a whole lot of other work.

lemire commented:
I'm confident we will make something out of this PR. We need to review carefully the performance issues and maybe do a pass of low-level optimizing.

michaeleisel commented:
agreed. also, a few things:

- i will revert back to when `compute_float_64` was faster (when strtod was the bottleneck) and investigate
- i should mention that how rapidjson parses most of the floats in canada.json is with `strtoddiyfp` in http://rapidjson.org/strtod_8h_source.html . from my benchmarks, it's a lot slower, but it's a little hard to compare apples-to-apples with it. i'd be curious to see what benchmarks against rapidjson in full precision mode reveal.
- we can do something very similar to `compute_float_64` but with 80-bit long doubles, and it seems substantially faster. however, i didn't initially include it because arm has 64-bit long doubles and there's no msvc support. but we could certainly add it (and even support it for msvc if we allow assembly files?)

michaeleisel commented:
by the way, by bare-metal linux, you just mean that it's not running in a vm?

michaeleisel commented:
also curious what strategies there are to reduce noise, e.g. if one linux os is better than another for consistent benchmarking

lemire commented:
> by bare-metal linux, you just mean that it's not running in a vm?

A VM is not bare-metal.

lemire commented:
> it's a little hard to compare apples-to-apples with it

What do you mean?

michaeleisel commented:
> A VM is not bare-metal.

Right, I'm just wondering if there are any further requirements for good benchmarking besides being on non-VM linux

> What do you mean?

rapidjson's strtoddiyfp function wants the doubles as strings without the ".", so "1.23" becomes "123". this means that, unless you modify the buffer, you have to `memmove` to get things right. you could mess with things to get a good comparison, i just haven't looked too closely at the code. i'd rather just benchmark against rapidjson at the end

michaeleisel commented:
i did do coimparisons, and they strongly suggest that rapidjson's is much slower regardless

lemire commented:
>  I'm just wondering if there are any further requirements for good benchmarking besides being on non-VM linux


As a hint.... the instruction count should always be the same... even on a noisy machine. For the kind of work we do in this project, it is almost always the case that reducing the instruction count translates into better performance. It is an imperfect, but easy-to-measure and reliable (meaning it does not change from run to run) metric.

This being said, if you want access to a test machine and you can do ssh, just email me at lemire@gmail.com.

michaeleisel commented:
my results are... strange. with 67f6c49, which makes things more like they were when perf was dominated by strtod, it gets faster on linux but slower on os x. although, we could at least contain this weirdness to arm if we use 80-bit long doubles (what are your thoughts on a .s assembly file in the code base, generated from C on a linux machine with a compiler? this would avoid msvc and its 64-bit long doubles)

lemire commented:
@michaeleisel 

Is the difference due to linux or to the standard library? What happens if you use GCC on both macOS and linux... or use clang on both macOS and linux?

If you think that the difference is due to the operating system (macOS/Linux) and not due to the compiler or standard library... what in the operating system do you think is at fault?

It seems unavoidable that if we have an expensive library call (strtod) on a critical path, then our performance will depend on the standard library with everything that it entails. That may not be desirable.

You refer to assembly... but if the assembly mostly calls the standard library, that does not help you very much...

michaeleisel commented:
i will address these issues. but first, i have a question: what are your thoughts on using `strtod` for all long (>= 19 digits) floats? i double-checked, and the max # of digits needed to represent any float is 17. so let's say a server is writing a float into json, it will only use that # of digits. this means that 19+ digits would only occur in rather unusual circumstances (i can't think of any i've seen personally). the benefit of this is that it makes the code simpler

lemire commented:
Sounds reasonable.

michaeleisel commented on include/simdjson/numberparsing.h:
```diff
@@ -569,20 +548,27 @@ static really_inline bool parse_number(const uint8_t *const buf, ParsedJson &pj,
-        // this is almost never going to get called!!!
-        // we start anew, going slowly!!!
-        return parse_float(buf, pj, offset, found_minus);
+        return parse_float_strtod(buf, pj, offset, p);
```
i just figured it's so rare, maybe we're fine with this for imprecise parsing as well

michaeleisel commented:
it's now at 21.71 cycles per char on that linux machine. it's pretty sensitive to seemingly small changes, like returning a double vs. returning a bool from the 3 functions (and using pointers to set the other var). identifying bottlenecks at this point is difficult for me, because apple's cpu profiler isn't working well, and i don't have much experience with linux sampling profilers. also, it turns out that long doubles actually seem to be slower, so i didn't include the long double function

lemire commented:
Ok. I can do my part.

Sampling is not good for this kind of work... or rather, it is limited. You get some identification of the bottlenecks but it is not nearly as useful as one might think.

You basically have to read the assembly...

michaeleisel commented:
i think we'll want to leave the _128 function alone for low-level optimization for right now, i'm thinking about some high-level optimizations for it

michaeleisel commented:
By the way, simdjson with full precision takes 3ms for canada.json on my machine, vs. 7 ms for rapidjson non-full-prec and 22ms(!) for rapidjson full-prec. so i think however the further optimizations turn out, it will be a big win. and while canada.json is in some sense silly, i think that it's par for the course for payloads that have arbitrary doubles

lemire commented:
I hear you.

michaeleisel commented:
hi, just checking in on this

lemire commented:
Still on my todo. This is important but also non-trivial.

lemire commented:
I am going to merge this PR into a local branch for further work.

lemire commented:
Closing in favour of https://github.com/lemire/simdjson/pull/333
